encoding/protobuf/buffer.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protobuf
+
+import (
+	"math"
+
+	"github.com/m3db/m3x/pool"
+)
+
+// Buffer contains a byte slice backed by an optional bytes pool.
+type Buffer struct {
+	buf    []byte
+	pool   pool.BytesPool
+	closed bool
+}
+
+// NewBuffer create a new buffer.
+func NewBuffer(buf []byte, p pool.BytesPool) Buffer {
+	return Buffer{buf: buf, pool: p}
+}
+
+// Bytes returns the raw byte slice.
+func (b *Buffer) Bytes() []byte { return b.buf }
+
+// Close closes the buffer.
+func (b *Buffer) Close() {
+	if b.closed {
+		return
+	}
+	b.closed = true
+	if b.pool != nil && b.buf != nil {
+		b.pool.Put(b.buf)
+	}
+	b.pool = nil
+	b.buf = nil
+}
+
+// allocate allocates a byte slice with at least the specified size.
+func allocate(p pool.BytesPool, targetSize int) []byte {
```

Comment:
nit: move this below `ensureBufferSize`?

Reply:
sure